kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,915 @@
+				if lbingress == addrs.String() {
+					skipAddrs = true
+				}
+			}
+
+			if skipAddrs {
+				continue
+			}
+
+			ipamArgs := ipam.AssignIPArgs{
+				IP:          addrs,
+				Hostname:    api.VirtualLoadBalancer,
+				HandleID:    &handle,
+				Attrs:       metadataAttrs,
+				IntendedUse: api.IPPoolAllowedUseLoadBalancer,
+			}
+
+			err = c.calicoClient.IPAM().AssignIP(context.Background(), ipamArgs)
+			if err != nil {
+				log.WithField("ip", addrs).WithError(err).Warn("failed to assign ip to node")
+				return nil, err
+			}
+			assignedIPs = append(assignedIPs, addrs.String())
+			c.allocationTracker.assignAddressToService(namespacedName, addrs.String())
+		}
+		return assignedIPs, err
+	}
+
+	// Build AssignArgs based on Service IP family attr
+	num4 := 0
+	num6 := 0
+
+	for _, ipFamily := range svc.Spec.IPFamilies {
+		if ipFamily == v1.IPv4Protocol {
+			num4++
+		}
+		if ipFamily == v1.IPv6Protocol {
+			num6++
+		}
+	}
+
+	// Check if IP from ipfamily is already assigned, skip it as we're trying to assign only the missing one.
+	// This can happen when error happened during the initial assignment, and now we're trying to assign ip again from the syncIPAM func
+	for ingress := range c.allocationTracker.ipsByService[namespacedName] {
+		if ip := cnet.ParseIP(ingress); ip != nil {
+			if ip.To4() != nil {
+				num4 = 0
+			}
+			if ip.To16() != nil {
+				num6 = 0
+			}
+		}
+	}
+
+	if num4 == 0 && num6 == 0 {
+		return nil, errors.New("No new IPs to assign, Service already has ipv4 and ipv6 address")
+	}
+
+	args := ipam.AutoAssignArgs{
+		Num4:        num4,
+		Num6:        num6,
+		IntendedUse: api.IPPoolAllowedUseLoadBalancer,
+		Hostname:    api.VirtualLoadBalancer,
+		HandleID:    &handle,
+		Attrs:       metadataAttrs,
+	}
+
+	if ipv4Pools != nil {
+		args.IPv4Pools = ipv4Pools
+	}
+
+	if ipv6Pools != nil {
+		args.IPv6Pools = ipv6Pools
+	}
+
+	v4Assignments, v6assignments, err := c.calicoClient.IPAM().AutoAssign(context.Background(), args)
+	if err != nil {
+		log.WithField("svc", svc.Name).WithError(err).Warn("error on assigning IP address to service")
+		return nil, err
+	}
+
+	if v4Assignments != nil {
+		for _, assignment := range v4Assignments.IPs {
+			assignedIPs = append(assignedIPs, assignment.IP.String())
+			c.allocationTracker.assignAddressToService(namespacedName, assignment.IP.String())
+		}
+	}
+
+	if v6assignments != nil {
+		for _, assignment := range v6assignments.IPs {
+			assignedIPs = append(assignedIPs, assignment.IP.String())
+			c.allocationTracker.assignAddressToService(namespacedName, assignment.IP.String())
+		}
+	}
+
+	return assignedIPs, nil
+}
+
+// releaseIPByHandle tries to release all IPs allocated with the Service unique handle
+func (c *loadBalancerController) releaseIPByHandle(namespacedName types.NamespacedName, handle string) error {
+	log.Info("Service type LoadBalancer deleted, releasing assigned IP address")
```

Comment:
This doesn't necessarily mean the service was deleted, right? It could be that it is no longer a LoadBalancer type?